File: src/gui/gui_x11.zig
//! gui_x11.zig
//! Module for interfacing with X11

const std = @import("std");
const clap = @cImport({
    @cInclude("clap/clap.h");
});
const c_cast = std.zig.c_translation.cast;
const Plugin = @import("../Plugin.zig");
const allocator = &Plugin.allocator;
pub const GUI_API = clap.CLAP_WINDOW_API_X11;
const Parent = @import("Gui.zig");
const GUI_WIDTH = Parent.GUI_WIDTH;
const GUI_HEIGHT = Parent.GUI_HEIGHT;
const pluginPaint = Parent.pluginPaint;

pub const x11 = @cImport({
    @cInclude("X11/Xlib.h");
    @cInclude("X11/Xutil.h");
    @cInclude("X11/Xatom.h");
});

pub const Display = struct {
    display: *x11.Display,
    window: x11.Window,
    image: *x11.XImage,
    bits: []u32,
};

pub fn GUICreate(plugin: *Plugin) !void {
    std.debug.assert(plugin.gui == null);
    plugin.gui = try allocator.create(Display);

    if (plugin.gui == null) return;

    plugin.gui.?.display = x11.XOpenDisplay(null).?;
    var attributes: x11.XSetWindowAttributes = undefined;
    plugin.gui.?.window = x11.XCreateWindow(plugin.gui.?.display, x11.DefaultRootWindow(plugin.gui.?.display), 0, 0, GUI_WIDTH, GUI_HEIGHT, 0, 0, x11.InputOutput, x11.CopyFromParent, x11.CWOverrideRedirect, &attributes);

    _ = x11.XStoreName(plugin.gui.?.display, plugin.gui.?.window, Plugin.PluginDesc.name);

    const embedInfoAtom = x11.XInternAtom(plugin.gui.?.display, "_XEMBED_INFO", 0);
    const embedInfoData = [2]u32{ 0, 0 };
    _ = x11.XChangeProperty(plugin.gui.?.display, plugin.gui.?.window, embedInfoAtom, embedInfoAtom, 32, x11.PropModeReplace, c_cast([*c]const u8, &embedInfoData), 2);

    var sizeHints: x11.XSizeHints = .{
        .flags = x11.PMinSize | x11.PMaxSize,
        .min_width = GUI_WIDTH,
        .max_width = GUI_WIDTH,
        .min_height = GUI_HEIGHT,
        .max_height = GUI_HEIGHT,
        .x = undefined,
        .y = undefined,
        .width = GUI_WIDTH,
        .height = GUI_HEIGHT,
        .width_inc = undefined,
        .height_inc = undefined,
        .min_aspect = undefined,
        .max_aspect = undefined,
        .base_width = undefined,
        .base_height = undefined,
        .win_gravity = undefined,
    };
    x11.XSetWMNormalHints(plugin.gui.?.display, plugin.gui.?.window, &sizeHints);

    _ = x11.XSelectInput(plugin.gui.?.display, plugin.gui.?.window, x11.SubstructureNotifyMask | x11.ExposureMask | x11.PointerMotionMask | x11.ButtonPressMask | x11.ButtonReleaseMask | x11.KeyPressMask | x11.KeyReleaseMask | x11.StructureNotifyMask | x11.EnterWindowMask | x11.LeaveWindowMask | x11.ButtonMotionMask | x11.KeymapStateMask | x11.FocusChangeMask | x11.PropertyChangeMask);

    plugin.gui.?.image = x11.XCreateImage(plugin.gui.?.display, x11.DefaultVisual(plugin.gui.?.display, 0), 24, x11.ZPixmap, 0, null, 10, 10, 32, 0);
    plugin.gui.?.bits = try allocator.alloc(u32, GUI_WIDTH * GUI_HEIGHT * 4);
    plugin.gui.?.image.width = GUI_WIDTH;
    plugin.gui.?.image.height = GUI_HEIGHT;
    plugin.gui.?.image.bytes_per_line = GUI_WIDTH * 4;
    plugin.gui.?.image.data = c_cast(*u8, plugin.gui.?.bits);

    if (plugin.host_posix_fd_support != null and plugin.host_posix_fd_support.*.register_fd != null) {
        _ = plugin.host_posix_fd_support.*.register_fd.?(plugin.host, x11.ConnectionNumber(plugin.gui.?.display), clap.CLAP_POSIX_FD_READ);
    }
}

pub fn GUIPaint(plugin: *Plugin, internal: bool) void {
    if (internal) pluginPaint(plugin.gui.?.bits.ptr);
    _ = x11.XPutImage(plugin.gui.?.display, plugin.gui.?.window, x11.DefaultGC(plugin.gui.?.display, 0), plugin.gui.?.image, 0, 0, 0, 0, GUI_WIDTH, GUI_HEIGHT);
}

fn GUIProcessX11Event(plugin: *Plugin, event: *x11.XEvent) void {
    if (event.type == x11.Expose) {
        if (event.xexpose.window == plugin.gui.?.window)
            GUIPaint(plugin, false);
    }
}

pub fn GUIOnPOSIXFD(plugin: *Plugin) void {
    _ = x11.XFlush(plugin.gui.?.display);

    const pending = x11.XPending(plugin.gui.?.display);
    std.debug.print("Pending: {d}\n", .{pending});

    if (pending > 0) {
        var event: x11.XEvent = undefined;
        _ = x11.XNextEvent(plugin.gui.?.display, &event);

        while (c_cast(bool, x11.XPending(plugin.gui.?.display))) {
            var event0: x11.XEvent = undefined;
            _ = x11.XNextEvent(plugin.gui.?.display, &event0);

            if (event.type == x11.MotionNotify and event0.type == x11.MotionNotify) {
                // merge adjacent mouse motion events
            } else {
                GUIProcessX11Event(plugin, &event);
                _ = x11.XFlush(plugin.gui.?.display);
            }

            event = event0;
        }

        GUIProcessX11Event(plugin, &event);
        _ = x11.XFlush(plugin.gui.?.display);
        GUIPaint(plugin, true);
    } else std.debug.print("No events!\n", .{});

    _ = x11.XFlush(plugin.gui.?.display);
}

pub fn GUIDestroy(plugin: *Plugin) void {
    std.debug.assert(plugin.gui != null);

    if (plugin.host_posix_fd_support != null and plugin.host_posix_fd_support.*.unregister_fd != null) {
        _ = plugin.host_posix_fd_support.*.unregister_fd.?(plugin.host, x11.ConnectionNumber(plugin.gui.?.display));
    }

    allocator.free(plugin.gui.?.bits);
    plugin.gui.?.image.data = null;
    _ = x11.XDestroyImage(plugin.gui.?.image);
    _ = x11.XDestroyWindow(plugin.gui.?.display, plugin.gui.?.window);
    _ = x11.XCloseDisplay(plugin.gui.?.display);

    allocator.destroy(plugin.gui.?);
    plugin.gui = null;
}

pub fn GUISetParent(plugin: *Plugin, clap_window: [*c]const clap.clap_window_t) void {
    _ = x11.XReparentWindow(plugin.gui.?.display, plugin.gui.?.window, c_cast(x11.Window, clap_window.*.unnamed_0.x11), 0, 0);
    _ = x11.XFlush(plugin.gui.?.display);
}

pub fn GUISetVisible(plugin: *Plugin, visible: bool) void {
    if (visible)
        _ = x11.XMapRaised(plugin.gui.?.display, plugin.gui.?.window)
    else
        _ = x11.XUnmapWindow(plugin.gui.?.display, plugin.gui.?.window);

    _ = x11.XFlush(plugin.gui.?.display);
}

File: src/gui/gui_x11.c
// X11 GUI impl.

#include <X11/X.h>
#include <X11/Xlib.h>
#include <X11/Xutil.h>
#include <X11/Xatom.h>

#include <stdlib.h>
#include <stdbool.h>
#include <stddef.h>
#include <stdint.h>
#include <assert.h>
#include <stdio.h>

typedef struct GuiImpl {
    Display *display;
    int fd;
    Window window;
    XImage *image;
    uint32_t *bits;
    uint32_t width, height;
    void *user;
} GuiImpl_t;

/// button 0 = dragging, 1 = click, -1 = release
extern void inputEvent(void* user, int32_t x, int32_t y, int8_t button);
extern void render(void*);

void guiRender(GuiImpl_t *gui, bool internal)
{
    if (internal) render(gui->user);
    XPutImage(gui->display, gui->window, DefaultGC(gui->display, 0), gui->image,
              0, 0, 0, 0, gui->width, gui->height);
}

GuiImpl_t *guiCreate(void *user, uint32_t *bits, uint32_t w, uint32_t h)
{
    GuiImpl_t *gui = (GuiImpl_t*)calloc(1, sizeof(GuiImpl_t));
    gui->user = user;
    gui->width = w;
    gui->height = h;
    gui->display = XOpenDisplay(NULL);
    XSetWindowAttributes attributes = {};
    gui->window = XCreateWindow(gui->display, DefaultRootWindow(gui->display),
                                0, 0, w, h, 0, 0, InputOutput, CopyFromParent,
                                CWOverrideRedirect, &attributes);
    XStoreName(gui->display, gui->window, "ZigVerb");

    Atom embedInfoAtom = XInternAtom(gui->display, "_XEMBED_INFO", 0);
    uint32_t embedInfoData[2] = {0, 0};
    XChangeProperty(gui->display, gui->window, embedInfoAtom, embedInfoAtom,
                    32, PropModeReplace, (uint8_t*)embedInfoData, 2);

    XSizeHints sizeHints = {};
    sizeHints.flags = PMinSize | PMaxSize;
    sizeHints.min_width = sizeHints.max_width = w;
    sizeHints.min_height = sizeHints.max_height = h;
    XSetWMNormalHints(gui->display, gui->window, &sizeHints);

    XSelectInput(gui->display, gui->window, SubstructureNotifyMask |
                ExposureMask | PointerMotionMask | ButtonPressMask |
                ButtonReleaseMask | KeyPressMask | KeyReleaseMask |
                EnterWindowMask | LeaveWindowMask | ButtonMotionMask |
                KeymapStateMask | FocusChangeMask | PropertyChangeMask);

    gui->image = XCreateImage(gui->display, DefaultVisual(gui->display, 0), 24,
                              ZPixmap, 0, NULL, 10, 10, 32, 0);
    gui->bits = bits;
    gui->image->width = w;
    gui->image->height = h;
    gui->image->bytes_per_line = w * 4;
    gui->image->data = (char *)gui->bits;
    gui->fd = ConnectionNumber(gui->display);

    return gui;
}

void processX11Event(GuiImpl_t *gui, XEvent *event)
{
    switch(event->type) {
    case Expose:
        if (event->xexpose.window == gui->window)
            guiRender(gui, true);
        break;
    case MotionNotify:
        if (event->xmotion.window == gui->window)
            inputEvent(gui->user, event->xmotion.x, event->xmotion.y, 0);
        break;
    case ButtonPress:
        if (event->xbutton.window == gui->window && event->xbutton.button == 1)
            inputEvent(gui->user, event->xbutton.x, event->xbutton.y, 1);
        break;
    case ButtonRelease:
        if (event->xbutton.window == gui->window && event->xbutton.button == 1)
            inputEvent(gui->user, event->xbutton.x, event->xbutton.y, -1);
        break;
    }
}

void guiOnPosixFd(GuiImpl_t *gui)
{
    XFlush(gui->display);

    if (XPending(gui->display)) {
        XEvent event;
        XNextEvent(gui->display, &event);
        while (XPending(gui->display)) {
            XEvent event0;
            XNextEvent(gui->display, &event0);
            if(event.type == MotionNotify && event0.type == MotionNotify) {
                // Merge adjacent mouse motion events
            } else {
                processX11Event(gui, &event);
                XFlush(gui->display);
            }
            event = event0;
        }
        processX11Event(gui, &event);
        XFlush(gui->display);
        guiRender(gui, true);
    }
}

void guiDestroy(GuiImpl_t *gui)
{
    assert(gui);

    gui->image->data = NULL;
    XDestroyImage(gui->image);
    XDestroyWindow(gui->display, gui->window);
    XCloseDisplay(gui->display);

    free(gui);
    gui = NULL;
}

void guiSetParent(GuiImpl_t *gui, Window parent_window)
{
    XReparentWindow(gui->display, gui->window, parent_window, 0, 0);
    XFlush(gui->display);
}

void guiSetVisible(GuiImpl_t *gui, bool visible)
{
    if (visible) XMapRaised(gui->display, gui->window);
    else XUnmapWindow(gui->display, gui->window);
    XFlush(gui->display);
}
